src/kbmod/search/raw_image_eigen.h

```diff
@@ -0,0 +1,250 @@
+  using Image = Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  using ImageI = Eigen::Matrix<int, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  using ImageRef = Eigen::Ref<Image>;
+  using ImageIRef = Eigen::Ref<Image>;
+
+
+  class RawImageEigen {
+  public:
+    RawImageEigen();
+    explicit RawImageEigen(Image& img, double obs_time=-1.0d);
+    explicit RawImageEigen(unsigned w, unsigned h, float value=0.0f, double obs_time=-1.0d);
+
+    RawImageEigen(const RawImageEigen& old);  // Copy constructor
+    RawImageEigen(RawImageEigen&& source);    // Move constructor
+
+    RawImageEigen& operator=(const RawImageEigen& source);  // Copy assignment
+    RawImageEigen& operator=(RawImageEigen&& source);       // Move assignment
+
+    // Basic getter functions for image data.
+    unsigned get_width() const { return width; }
+    unsigned get_height() const { return height; }
+    unsigned get_npixels() const { return width * height; }
+
+    // Inline pixel functions.
+    float get_pixel(int x, int y) const {
+      return Index(x, y).is_within(width, height) ? image(x, y) : NO_DATA;
+    }
+
+    bool pixel_has_data(int x, int y) const {
+      return Index(x, y).is_within(width, height) ? image(x, y) != NO_DATA : false;
+    }
+
+    void set_pixel(int x, int y, float value) {
+      if (Index(x, y).is_within(width, height))
+        image(y, x) = value;
+    }
+
+    // Functions for locally storing the image time.
+    double get_obstime() const { return obstime; }
+    void set_obstime(double new_time) { obstime = new_time; }
+
+    // note, this produces a copy, because array is private
+    // it is not possible to return an usable reference to it.
+    const Image& get_array() const { return image; }
+    // this will be a raw pointer to the underlying array
+    // we use this to copy to GPU and nowhere else!
+    float* data() { return image.data(); }
+    void set_all(float value);
+
+
+    // Check if two raw images are approximately equal.
+    bool isclose(const RawImageEigen& imgB, float atol) const;
+
+    // Get the interpolated brightness of a real values point
+    // using the four neighboring array.
+    auto get_interp_neighbors_and_weights(const float x, const float y) const;
+    float interpolate(const float x, const float y) const;
+    float interpolate2(const float x, const float y) const;
+
+    // Create a "stamp" image of a give radius (width=2*radius+1) about the
+    // given point.
+    // keep_no_data indicates whether to use the NO_DATA flag or replace with 0.0.
+    RawImageEigen create_stamp(const float x, float y,
+                               const int radius, const bool interpolate,
+                               const bool keep_no_data) const;
+
+    // pixel modifiers
+    void add(const float x, const float y, const float value);
+    void add(const unsigned x, const unsigned y, const float value);
+    void add(const Index p, const float value);
```

Comment:
Do we need all these versions? Can we rely on implicit casting?

Reply:
As mentioned above, fast and loose with casting makes for mistakes. I'm going to close this in favor of continuing the discussion in this thread: https://github.com/dirac-institute/kbmod/pull/364#discussion_r1351706890